lisa/_assets/kmodules/lisa/fs.c

```diff
@@ -0,0 +1,755 @@
+
+static struct dentry *
+lisa_fs_create_single(struct dentry *parent, const char *name,
+		      const struct inode_operations *i_ops,
+		      const struct file_operations *f_ops, umode_t mode,
+		      void *data)
+{
+	struct dentry *dentry;
+	struct inode *inode;
+
+	dentry = d_alloc_name(parent, name);
+	if (!dentry)
+		return NULL;
+	inode = lisa_fs_create_inode(parent->d_sb, mode);
+	if (!inode) {
+		dput(dentry);
+		return NULL;
+	}
+
+	if (mode & S_IFREG) {
+		inode->i_fop = f_ops;
+	} else {
+		inode->i_op = i_ops;
+		inode->i_fop = f_ops;
+	}
+	inode->i_private = data;
+	d_add(dentry, inode);
+	if (mode & S_IFDIR) {
+		inc_nlink(d_inode(parent));
+		inc_nlink(inode);
+	}
+
+	return dentry;
+}
+
+/* Called with interface_lock */
+static int
+lisa_fs_create_files(struct dentry *parent, bool is_top_level, struct lisa_cfg *cfg)
+{
+	struct feature_param_entry *entry;
+	struct feature *feature;
+
+	entry = allocate_feature_param_entry();
+	if (!entry)
+		return -ENOMEM;
+
+	init_feature_param_entry(entry, cfg, &lisa_features_param);
+
+	/* set_features: enable a feature - RW. */
+	if (!lisa_fs_create_single(parent, "set_features",
+			      NULL,
+			      &lisa_param_feature_fops,
+			      S_IFREG | S_IRUGO | S_IWUGO, entry)) {
+		free_feature_param_entry(entry);
+		return -ENOMEM;
+	}
+
+	/* available_features: list available features - RO. */
+	if (!lisa_fs_create_single(parent, "available_features",
+			      NULL,
+			      &lisa_available_features_fops,
+			      S_IFREG | S_IRUGO, &lisa_features_param))
+		return -ENOMEM;
+
+	/* activate: activate the selected (and configured) features - RW. */
+	if (!lisa_fs_create_single(parent, "activate",
+			      NULL,
+			      &lisa_activate_fops,
+			      S_IFREG | S_IRUGO | S_IWUGO, cfg))
+		return -ENOMEM;
+
+	/* configs: Dir containing configurations, only setup at the top level. */
+	if (is_top_level) {
+		if (!lisa_fs_create_single(parent, "configs",
+				&lisa_fs_dir_inode_operations,
+				&simple_dir_operations,
+				S_IFDIR | S_IRUGO, NULL))
+			return -ENOMEM;
+	}
+
+	/* Create a dir for features having parameters. */
+	for_each_feature(feature) {
+		struct feature_param *param, **pparam;
+		struct dentry *dentry;
+
+		if (!feature->params)
+			continue;
+
+		dentry = lisa_fs_create_single(parent, feature->name,
+					       &simple_dir_inode_operations,
+					       &simple_dir_operations,
+					       S_IFDIR | S_IRUGO, cfg);
+		if (!dentry) {
+			pr_err("Failed to initialize feature's (%s) root node\n",
+			       feature->name);
+			return -ENOMEM;
+		}
+
+		for_each_feature_param(param, pparam, feature) {
+			entry = allocate_feature_param_entry();
+			if (!entry)
+				return -ENOMEM;
+
+			init_feature_param_entry(entry, cfg, param);
+
+			if (!lisa_fs_create_single(dentry, param->name, NULL,
+						   &lisa_param_feature_fops,
+						   S_IFREG | S_IRUGO, entry)) {
+				free_feature_param_entry(entry);
+				return -ENOMEM;
+			}
+		}
+	}
```

Comment:
Will all the dentry created on the way before an error is triggered be clean-up ? I do  not recall how fs framework handles such cases (as the fs_fill_super will fail in this case)
Also, same question about allocated entries - will those be covered as well and memory release ?
(aside : did you try to run the module with kmemleak being enabled - it should show if smth falls between the crack )